client/tinychain/ml/linalg.py

```diff
@@ -269,7 +273,467 @@ def step(i: UInt):
+        U_final = After([update_prev_col, update_col], U)
+        return If(cxt.f.abs() <= EPS, 
+                Map(running=cxt.f.abs() <= EPS, i=i + 1, U=U_final, W=W, e=e),
+                Map(running=cxt.f.abs() <= EPS, i=i + 1, U=U, W=W, e=e))
+
+    @closure(k)
+    @post_op
+    def cond(running: Bool, i: UInt) -> Bool:
+        return running.logical_and(i < k + 1)
+
+    return Map(While(cond, step, Map(running = True, i=l)))    
+
+
+@post_op
+def testfsplit(txn, U, W, e, k):
+    @closure(U, W, e, k)
+    @post_op
+    def step(cxt, l: UInt) -> Map:
+        goto_test_f_convergence = (e[l].abs() <= EPS)
+        running = W[l-1].abs() > EPS
+        return Map(l=l - 1, goto_test_f_convergence=goto_test_f_convergence, running=running, U=U, W=W, e=e)
+
+    @closure(k)
+    @post_op
+    def cond(l: UInt, goto_test_f_convergence:Bool, running: Bool) -> Bool:
+        running.logical_and(goto_test_f_convergence.logical_not()).logical_and(l >= 0)
+
+    txn.res = Map(While(cond, step, Map(l=k, goto_test_f_convergence = False)))
+    txn.cancelation = cancelation(U, W, e, txn.l, k)
+
+    return If(txn.res['goto_test_f_convergence'], txn.res['l'], txn.cancelation)
+
+
+@post_op
+def golub_kahan(txn, U: Tensor, W: Tensor, V: Tensor, e: Tensor, k: UInt, maxiter=30):
+    """
+    Diagonalization of the bidiagonal form: 
+        - k is the kth singular value
+        - loop over maxiter allowed iteration
+    """
+
+    @closure(U, W, e)
+    @post_op
+    def step(cxt, t: UInt, k: UInt, maxiter: UInt) -> Map:
+        
+        cxt.l = testfsplit(U, W, e, k)
+        cxt.update_mtrx = If(W[k] < 0.0, [W[k].write(W[k].abs()), V[:,k].write(V[:,k].abs())], W)
+        W_upd_init = If(cxt.l != k, W, cxt.update_mtrx)
+        running = cxt.l != k
+        convergence = If(t == maxiter-1, 
+                        BadRequest(String("SVD Error: no convergence found. t is {{t}}").render(t=t)),
+                        True)
+
+        # shift from bottom 2x2 minor
+        x = W[cxt.l]
+        y = W[k-1]
+        z = W[k]
+        g = e[k-1]
+        h = e[k]
+
+        f = ((y-z)*(y+z)+(g-h)*(g+h))/(2.0*h*y)
+        g = pythag(f, 1.0)
+        f = ((x-z)*(x+z)+h*((y/(f+sign_like(g,f)))-h))/x
+
+        # next QR transformation
+        c = 1.0
+        s = 1.0
+
+        @closure(U, W, V, e)
+        @post_op
+        def step(cxt, i: UInt, x: F32, c: F32, s: F32, f: F32) -> Map:
+            g= e[i]
+            y = W[i]
+            h = s*g
+            g = c*g
+
+            z = pythag(f,h)
+            e[i-1] = z
+
+            c = f/z
+            s = h/z
+            f = x*c+g*s
+            g = g*c-x*s
+            h = y*s
+            y = y*c
+
+            cxt.X = V[:,i-1].copy()
+            cxt.Z = V[:,i].copy()
+            V_upd = After([cxt.X, cxt.Z], V[:, i-1].write(c * cxt.X + s * cxt.Z))
+            V_upd_final = After(V_upd, V[:, i].write(c * cxt.Z - s * cxt.X))
+
+            z = pythag(f, h)
+            W[i - 1].write(z)
```

Comment:
Actually, it doesn't look like this write will execute because there's nothing that explicitly depends on it--do you need to `return After(w[i - 1].write(z), ...)`?